pkg/datapath/loader/loader.go

```diff
@@ -288,24 +289,27 @@ func (l *Loader) reloadDatapath(ctx context.Context, ep datapath.Endpoint, dirs
 			return err
 		}
 	} else {
-		finalize, err := replaceDatapath(ctx, ep.InterfaceName(), objPath, symbolFromEndpoint, dirIngress, "")
-		if err != nil {
-			scopedLog := ep.Logger(Subsystem).WithFields(logrus.Fields{
-				logfields.Path: objPath,
-				logfields.Veth: ep.InterfaceName(),
-			})
-			// Don't log an error here if the context was canceled or timed out;
-			// this log message should only represent failures with respect to
-			// loading the program.
-			if ctx.Err() == nil {
-				scopedLog.WithError(err).Warn("JoinEP: Failed to attach ingress program")
+		var netNS ns.NetNS
+		var err error
+		if ep.NetNS() != "" {
+			netNS, err = ns.GetNS(ep.NetNS())
+			if err != nil {
+				return fmt.Errorf("failed to open netns %q: %s", ep.NetNS(), err)
+			}
+		} else {
+			netNS, err = ns.GetCurrentNS()
+			if err != nil {
+				return fmt.Errorf("failed to open current netns: %s", err)
 			}
-			return err
 		}
-		defer finalize()
-
-		if ep.RequireEgressProg() {
-			finalize, err := replaceDatapath(ctx, ep.InterfaceName(), objPath, symbolToEndpoint, dirEgress, "")
+		defer netNS.Close()
+		return netNS.Do(func(_ ns.NetNS) error {
```

Comment:
This is a big change - moving the BPF program from the host to the container namespace. What sort of effects will this have? What about existing pods on upgrade?

Reply:
If we didn't use `generic-device` chaining plugin before, netNS is host netNS，it does't affect existing pods on upgrade.
It only works for pods that use `generic-device` chaining plugin.

Reply:
Why can't you move the device from the CNI? It's very similar to what we had in the ipvlan mode which was removed by https://github.com/cilium/cilium/pull/20453. Also, I see some similarities with https://github.com/cilium/cilium/pull/14436.